rpcclient/roundrobin.go

```diff
@@ -0,0 +1,434 @@
+
+	// ConnectionState gets the connectionstate of corresponding wsConn.
+	ConnectionState(host string) (state ConnectionState, ok bool)
+
+	// WsConnection gets the wsConn corresponding to the host.
+	WsConnection(host string) (wsConn *websocket.Conn, connOk bool)
+
+	// NextDisconnectedWsConn will iterate over the connection state map and
+	// return the first wsConnection that has its state as Disconnected.
+	NextDisconnectedWsConn() *HostAddress
+
+	// NotifyConnStateChange is called by rpcClient when the ConnectionState
+	// for a hostaddress changes.
+	NotifyConnStateChange(sc *HostAddress, state ConnectionState, sync bool)
+
+	// UpdateReconnectAttempt will increase retryattempt + 1.
+	UpdateReconnectAttempt(hostAdd *HostAddress)
+
+	// NotifyReconnect will update the map for wsConns and update the connection state
+	// for this address.
+	NotifyReconnect(wsConn *websocket.Conn, hostAdd *HostAddress)
+
+	// Close closes the passed ws connection and sets the state as Disconnected
+	// by calling NotifyConnStateChange.
+	// Returns true if this was the last WS connection to be closed.
+	Close(wsConn *websocket.Conn) bool
+
+	// CloseAll closes all the ws connections which are active at present.
+	CloseAll()
+}
+
+// Picker is used to pick the next connection to be used.
+type Picker interface {
+	// Pick checks for the next usable connection using the connState map with balancer.
+	Pick(balancer Balancer) (*HostAddress, error)
+}
+
+type picker struct {
+	conns []HostAddress
+	mu    sync.Mutex
+	next  int
+}
+
+// Pick checks for the next usable connection using the connState map with the balancer.
+// It changes the connectionstate to Connecting if the Host picked is not yet used.
+// This would error out if there are no HostAddresses available for round-robin.
+func (p *picker) Pick(balancer Balancer) (conn *HostAddress, err error) {
+
+	if len(p.conns) <= 0 {
+		log.Infof(ErrNoConnAvailable.Error())
+		return conn, ErrNoConnAvailable
+	}
+
+	p.mu.Lock()
+	startPos := p.next
+	var ok bool
+
+	for {
+		conn = &p.conns[p.next]
+		var state ConnectionState
+		state, ok = balancer.ConnectionState(conn.Host)
+		p.next = (p.next + 1) % len(p.conns)
+		if !ok || state == ready || state == idle {
+			// This Host is either not yet tried or its ready to use.
+			break
+		}
+		// This indicates iteration is complete.
+		if startPos == p.next {
+			break
+		}
+	}
+	if conn == nil {
+		return conn, ErrNoUsableConnAvailable
+	}
+	if !ok {
+		balancer.NotifyConnStateChange(conn, connecting, true)
+	}
+	p.mu.Unlock()
+	return conn, nil
+}
+
+// WsInHandlerForConn represents the incoming message handler for a
+// ws connection. This is invoked as a goroutine, one per ws connection
+type WsInHandlerForConn func(*websocket.Conn, string)
+
+// RoundRobinBalancer represents a round-robin balancer. It implements Balancer
+// interface and maintains a Picker for picking Hosts in a round-robin manner.
+type RoundRobinBalancer struct {
```

Comment:
This should not be exported.  It is not directly usable by callers outside of the `rpcclient` package and it pollutes the public API by being exported.